File: robot-factory/csharp/src/RobotFactory/RobotOrder.cs
namespace RobotFactory;

public class RobotOrder
{
    private static readonly PartType[] AllPartTypes =
        Enum.GetValues<PartType>();

    private readonly Dictionary<PartType, PartOption> _parts = new();

    public IReadOnlyDictionary<PartType, PartOption> Parts => _parts;

    public void Configure(PartType type, PartOption option)
    {
        _parts[type] = option;
    }

    public void Validate()
    {
        var missing = AllPartTypes.Where(t => !_parts.ContainsKey(t)).ToList();
        if (missing.Count > 0)
        {
            throw new OrderIncompleteException(
                $"Order is missing part types: {string.Join(", ", missing)}");
        }
    }
}

File: robot-factory/typescript/src/RobotOrder.ts
import { OrderIncompleteError } from './OrderIncompleteError.js';
import type { PartOption } from './PartOption.js';
import { AllPartTypes, type PartType } from './PartType.js';

export class RobotOrder {
  private readonly _parts = new Map<PartType, PartOption>();

  get parts(): ReadonlyMap<PartType, PartOption> { return this._parts; }

  configure(type: PartType, option: PartOption): void {
    this._parts.set(type, option);
  }

  validate(): void {
    const missing = AllPartTypes.filter(t => !this._parts.has(t));
    if (missing.length > 0) {
      throw new OrderIncompleteError(
        `Order is missing part types: ${missing.join(', ')}`);
    }
  }
}
